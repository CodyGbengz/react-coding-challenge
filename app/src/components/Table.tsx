import React from 'react';
import styled from 'styled-components';

import useTheme from '../hooks/useTheme';

const StyledTable = styled.div`
    table {
        border-collapse: collapse;
        color: ${(props) => props.theme.text};
        width: 100%;
    }
    
    td, th {
        border: 1px solid #999;
        padding: 0.5rem;
        text-align: left;
    }
     
    .clickableHeader: hover {
        cursor: pointer;
    }

    tr:nth-child(odd) {
        background: ${(props) => props.theme.secondary};
    }

    tr:nth-child(even) {
        background: ${(props) => props.theme.primary};
    }
    
    img {
        border-radius: 50%;
        height: 80px;
        object-fit: cover;
        width: 80px;
    }

    img: hover {
        cursor: pointer;
    }
`;

const StyledSpan = styled.span`
    background: ${props => { 
        if(props.color === 'new') {
            return '#00f';
        }

        if(props.color === 'processing') {
            return '#ff0';
        }

        if(props.color === 'done') {
            return '#0f0';
        }

        if(props.color === 'error') {
            return '#f00'
        }
    }};
    border: 1px solid black;
    border-radius: 5px;
    display: inline-block;
    padding: 5px;
    text-align: center;
    width: 100px;
`;

export interface TableProps {
    items: IItem[];
    sortDirection: string;
    sortItems: () => void;
    selectItem: (id: number) => void;
    filterTerm: string;
}

export const Table: React.FC<TableProps> = ({
    items,
    sortDirection,
    sortItems,
    selectItem,
    filterTerm,
}) => {
    const filteredItems = items?.filter((item: IItem) => item.name.toLowerCase().indexOf(filterTerm) > -1);
    const listData = filterTerm ? filteredItems : items;
    const { theme } = useTheme();

    return (
        <StyledTable theme={theme}>
            <table>
                <thead>
                    <tr>
                        <th
                          onClick={() => sortItems()}
                          className="clickableHeader"
                          data-testid="name-column"
                        >
                          Name {sortDirection === 'asc' ? <span>&#8615;</span> : <span>&#8613;</span> }
                        </th>
                        <th>Image</th>
                        <th>Status</th>
                    </tr>
                </thead>
                <tbody>
                    {listData?.map(item => (
                    <tr key={item.id}>
                        <td >
                            {item.name}
                        </td>
                        <td >
                            <img
                                src={item.image || 'http://via.placeholder.com/300'}
                                alt="item image"
                                onClick={() => selectItem(item.id)}
                                data-testid={item.name}
                                />

                        </td>
                        <td >
                            <StyledSpan color={item.status}>{item.status}</StyledSpan>
                        </td>
                    </tr>
                    ))}
                </tbody>
            </table>
        </StyledTable>
    )
}
