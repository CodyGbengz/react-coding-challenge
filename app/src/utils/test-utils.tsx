import React, {FC, ReactElement} from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

import  theme  from '../configs/theme';
import { ThemeContext } from '../Root';

const AllTheProviders: FC = ({children}) => {
  return (
    <ThemeProvider theme={theme}>
        <ThemeContext.Provider value={{ theme }}>
            {children}
        </ThemeContext.Provider>
    </ThemeProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, {wrapper: AllTheProviders, ...options})

export * from '@testing-library/react'
export {customRender as render}