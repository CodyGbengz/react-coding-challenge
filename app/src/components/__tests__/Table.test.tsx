import React from 'react';
import { screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom'

import { render } from '../../utils/test-utils'
import { Table } from '../Table';

describe('Table', () => {
    const items:IItem[] = [
        {
            id: 2,
            name: "Dragon",
            image: "https://picsum.photos/200/300",
            status: "new"
        },
        {
            id: 1,
            name: "Dog",
            image: "https://picsum.photos/200/300",
            status: "error"
        },
        
    ];
    const selectItem = jest.fn();
    const sortItems = jest.fn()
    it('should render correctly', () => {
        const { container } = render(
            <Table
                items={items}
                sortDirection='asc'
                selectItem={selectItem}
                filterTerm=''
                sortItems={sortItems}
            />
        )
        expect(screen.getByText('Dragon')).toBeInTheDocument();
        expect(container.parentElement).toMatchSnapshot();
    });


    it('should call selectItem when item image is clicked', () => {
        render(<Table
            items={items}
            sortDirection='asc'
            selectItem={selectItem}
            filterTerm=''
            sortItems={sortItems}
        />
        )
        let itemImage = screen.getByTestId('Dragon');
        fireEvent.click(itemImage);
        expect(selectItem).toHaveBeenCalled()
    });

    it('should call sortItem when Name column is clicked', () => {
        render(<Table
            items={items}
            sortDirection='asc'
            selectItem={selectItem}
            filterTerm=''
            sortItems={sortItems}
        />
        )
        let nameColumn = screen.getByTestId('name-column');
        fireEvent.click(nameColumn);
        expect(sortItems).toHaveBeenCalled()
    });
})
