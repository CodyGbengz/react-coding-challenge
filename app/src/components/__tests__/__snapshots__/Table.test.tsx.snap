// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render correctly 1`] = `
<body>
  <div>
    <div
      class="sc-bdvvtL juZyqr"
    >
      <table>
        <thead>
          <tr>
            <th
              class="clickableHeader"
              data-testid="name-column"
            >
              Name 
              <span>
                â†§
              </span>
            </th>
            <th>
              Image
            </th>
            <th>
              Status
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Dragon
            </td>
            <td>
              <img
                alt="item image"
                data-testid="Dragon"
                src="https://picsum.photos/200/300"
              />
            </td>
            <td>
              <span
                class="sc-gsDKAQ jgQYRQ"
                color="new"
              >
                new
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Dog
            </td>
            <td>
              <img
                alt="item image"
                data-testid="Dog"
                src="https://picsum.photos/200/300"
              />
            </td>
            <td>
              <span
                class="sc-gsDKAQ eykIlQ"
                color="error"
              >
                error
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
